lib/persistence/abstract.js

```diff
@@ -52,6 +52,12 @@ AbstractPersistence.prototype.wire = function(server) {
   server.persistence = this;
 
   server.storePacket = function(packet, cb) {
+
+    // store qos 0 packets only if store_qos0_messages is true
+    if(packet.qos === 0 && packet.retain === false && ! that.options.store_qos0_messages){
+      return cb();
+    }
+
```

Comment:
you should move this block later in the fuction, otherwise retained messages with QoS 0 will not be stored. It should be after: https://github.com/mazhack/mosca/blob/master/lib/persistence/abstract.js#L67-L70